src/Data/Type.hs

```diff
@@ -1,51 +1,81 @@
-  F.Free (TVar name)        -> Set.singleton name
-  F.Free (ForAll name body) -> Set.delete name body
-  _                         -> fold ty
-
-returnType :: Free Type a -> Maybe (Free Type a)
-returnType (Free (_ :-> returnTy)) = Just returnTy
-returnType (Pure err)              = Just (Pure err)
-returnType _                       = Nothing
-
-fstType :: Free Type a -> Maybe (Free Type a)
-fstType (Free (fstTy :* _)) = Just fstTy
-fstType (Pure err)          = Just (Pure err)
-fstType _                   = Nothing
-
-sndType :: Free Type a -> Maybe (Free Type a)
-sndType (Free (_ :* sndTy)) = Just sndTy
-sndType (Pure err)          = Just (Pure err)
-sndType _                   = Nothing
-
-
-instance Binder (Free Type a) where
-  substitute = flip (cata (\ ty subst -> case ty of
-    F.Free (TVar name)        -> fromMaybe (Free (TVar name)) (substLookup name subst)
-    F.Free (ForAll name body) -> Free (ForAll name (body (substDelete name subst)))
-    F.Free other              -> Free (($ subst) <$> other)
-    F.Pure err                -> Pure err))
+instance Eq1 Type where liftEq = genericLiftEq
+instance Ord1 Type where liftCompare = genericLiftCompare
+instance Show1 Type where liftShowsPrec = genericLiftShowsPrec
+
+type PartialType = Free Type
+type TotalType = Fix Type
+
+totalToPartial :: TotalType -> PartialType a
+totalToPartial = cata wrap
+
+partialToTotal :: PartialType a -> Either [a] TotalType
+partialToTotal = iter (fmap Fix . sequenceA) . fmap (Left . pure)
+
+
+tvar :: Name -> PartialType a
+tvar name = wrap (TVar name)
+
+makeForAll :: Name -> PartialType a -> PartialType a
+makeForAll name body = wrap (ForAll name body)
+
+(.->) :: PartialType a -> PartialType a -> PartialType a
+arg .-> ret = wrap (arg :-> ret)
+
+infixr 0 .->
+
+(.*) :: PartialType a -> PartialType a -> PartialType a
+fst .* snd = wrap (fst :* snd)
+
+infixl 7 .*
+
+bool :: PartialType a
+bool = wrap Bool
+
+
+class FreeTypeVariables t where
+  freeTypeVariables :: t -> Set.Set Name
+
+instance FreeTypeVariables (PartialType a) where
+  freeTypeVariables = iter freeTypeVariables . (Set.empty <$)
+
+instance FreeTypeVariables t => FreeTypeVariables (Type t) where
+  freeTypeVariables (TVar name)        = Set.singleton name
+  freeTypeVariables (ForAll name body) = Set.delete name (freeTypeVariables body)
+  freeTypeVariables ty                 = foldMap freeTypeVariables ty
+
+instance FreeTypeVariables (Set.Set Name) where
+  freeTypeVariables = id
+
+
+substType :: Binder a a => Subst a -> Type a -> Either (Type a) a
+substType subst (TVar name)        = maybe (Left (TVar name)) Right (substLookup name subst)
+substType subst (ForAll name body) = Left (ForAll name (substitute (substDelete name subst) body))
+substType subst (arg :-> ret)      = Left (substitute subst arg :-> substitute subst ret)
+substType subst (fst :* snd)       = Left (substitute subst fst :* substitute subst snd)
+substType _     Bool               = Left Bool
```

Comment:
`Binder` doesn’t allow the value you’re substituting within and the result to have different types, but we need this to in order to apply substitutions of type variables (which will be of type `a` rather than `Type a`).